core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+    }
+}
+
+#[allow(clippy::expect_used)]
+impl P2PSystem {
+    #[inline]
+    /// Constructor for [`Self`]
+    pub fn new(listen_addr: String, public_key: PublicKey) -> P2PSystem {
+        P2PSystem {
+            listen_addr,
+            public_key,
+            connect_peer_target: Mutex::new(Vec::new()),
+            connected_to_peers: Mutex::new(HashMap::new()),
+            read_thread_data: Mutex::new(ReadThreadData {
+                poll_network_index: 0_u32,
+                packet_cache: Vec::new(),
+            }),
+        }
+    }
+
+    #[allow(clippy::unwrap_in_result)]
+    fn try_resolve_address(&self) -> Option<std::net::SocketAddr> {
+        let connected_to_peer_keys = {
+            let mut keys = Vec::new();
+            let mut connected_to_peers = self
+                .connected_to_peers
+                .lock();
+            for (public_key, _) in connected_to_peers.iter_mut() {
+                keys.push(public_key.clone());
+            }
+            keys
+        };
+        let target_addrs: Vec<String> = self
+            .connect_peer_target
+            .lock()
+            .iter()
+            .filter(|peer_id| {
+                !connected_to_peer_keys
+                    .iter()
+                    .any(|key| key == &peer_id.public_key)
+            })
+            .map(|peer_id| peer_id.address.clone())
+            .collect();
+        if target_addrs.is_empty() {
+            None
+        } else {
+            let address_candidate = &target_addrs[rand::random::<usize>() % target_addrs.len()];
+            // to_socket_addrs enables dns look-ups.
+            let maybe_addr = address_candidate
+                .to_socket_addrs()
+                .unwrap_or_else(|error| {
+                    debug!(?error, "Socket addr");
+                    Vec::new().into_iter()
+                })
+                .next();
+            if maybe_addr.is_none() {
+                error!(%address_candidate, "Error can't produce address from str. ");
+            }
+            maybe_addr
+        }
+    }
+
+    pub(crate) fn post_to_network(&self, message: &NetworkMessage, recipients: &[PublicKey]) {
+        let mut to_disconnect_keys = Vec::new();
+
+        let mut connected_to_peers = self
+            .connected_to_peers
+            .lock();
+        for public_key in recipients.iter() {
+            if let Some(PeerConnection { stream, crypto, .. }) =
+                connected_to_peers.get_mut(public_key)
+            {
+                let encrypted = crypto
+                    .encrypt(message.encode().clone())
+                    .expect("We should always be able to encrypt.");
+                let write1 = stream.write_all(&(u32::try_from(encrypted.len())).expect("Encrypted length exceeds 32 bits. This will cause issues with Parity Scale codec. Aborting. ").to_le_bytes());
+                let write2 = stream.write_all(&encrypted);
+
+                if write1.is_err() || write2.is_err() {
```

Comment:
This should be refactored. We have a good use for the `?` operator. 

This block is a function that we can neatly add to a `map` on `connected_to_peers.get_mut`, pipelined from `recipients.iter()` and collected into `to_disconnect_keys`. 

I'm not sure if we should remove the `expect`, or match on the returned result. We **do** have an `Error::Keys(CryptographicError::Encrypt(err))` variant, though I doubt we can meaningfully handle it inside an `iter()` block.